File: scripts/bench-enum-match.mjs
import { performance } from "node:perf_hooks";
import { newTag } from "../lib/calcit.procs.mjs";

const variantCount = 16;
const iterations = 5_000_000;
const tags = Array.from({ length: variantCount }, (_, idx) => newTag(`enum-match-${idx.toString().padStart(2, "0")}`));
const target = tags.at(-1);

const ifBody = tags.map((tag, idx) => `${idx === 0 ? "if" : "else if"} (tag === tags[${idx}]) return ${idx};`).join("\n");
const switchBody = tags.map((tag, idx) => `case ${tag.idx}: return ${idx};`).join("\n");
const linearDispatch = new Function("tag", "tags", `${ifBody}\nreturn -1;`);
const switchDispatch = new Function("tag", `switch (tag.idx) {${switchBody}} return -1;`);

run("if-else-tag-chain", () => linearDispatch(target, tags));
run("integer-tag-switch", () => switchDispatch(target));

function run(label, dispatch) {
  let checksum = 0;
  for (let idx = 0; idx < 100_000; idx += 1) checksum += dispatch();
  const started = performance.now();
  for (let idx = 0; idx < iterations; idx += 1) checksum += dispatch();
  const elapsed = performance.now() - started;
  console.log(`${label}: ${elapsed.toFixed(2)}ms checksum=${checksum}`);
}

File: examples/bench_enum_match.rs
use std::hint::black_box;
use std::sync::Arc;
use std::time::Instant;

use calcit::{Calcit, CalcitEnumDef, CalcitList, CalcitStructDef, CalcitStructValue};
use cirru_edn::EdnTag;

const ITERATIONS: usize = 5_000_000;
const VARIANT_COUNT: usize = 16;

fn main() {
  let fields: Vec<EdnTag> = (0..VARIANT_COUNT).map(|idx| EdnTag::new(format!("variant-{idx:02}"))).collect();
  let values = vec![Calcit::from(CalcitList::Vector(vec![])); fields.len()];
  let enum_def = CalcitEnumDef::from_struct(CalcitStructValue {
    struct_ref: Arc::new(CalcitStructDef::from_fields(EdnTag::new("DispatchState"), fields)),
    values: Arc::new(values),
  })
  .expect("valid benchmark enum");
  let target = enum_def.variants().last().expect("last variant").tag.clone();

  run("linear-tag-scan", || {
    enum_def
      .variants()
      .iter()
      .position(|variant| variant.tag == target)
      .expect("target variant")
  });
  run("indexed-branch-slot", || enum_def.variant_index(&target).expect("target variant"));
}

fn run(label: &str, mut dispatch: impl FnMut() -> usize) {
  let mut checksum = 0usize;
  for _ in 0..100_000 {
    checksum = checksum.wrapping_add(black_box(dispatch()));
  }
  let started = Instant::now();
  for _ in 0..ITERATIONS {
    checksum = checksum.wrapping_add(black_box(dispatch()));
  }
  println!("{label}: {:.2}ms checksum={checksum}", started.elapsed().as_secs_f64() * 1000.0);
}
